storage/replica.go

```diff
@@ -1151,6 +1156,97 @@ func (r *Replica) maybeSetCorrupt(err error) error {
 	return err
 }
 
+// resolveIntents resolves the given intents. Those which are local to the range
+// use a fast-path by submitting directly to the range-local Raft instance; all
+// others are dispatched in a batch.
+// TODO(tschottdorf): once Txn records have a list of possibly open intents,
+// resolveIntents should send an RPC to update the transaction(s) as well (for
+// those intents with non-pending Txns).
+// TODO(tschottdorf): need tests for this with range intents.
+func (r *Replica) resolveIntents(ctx context.Context, intents []proto.Intent) {
+	trace := tracer.FromCtx(ctx)
+	tracer.ToCtx(ctx, nil) // we're doing async stuff below; those need new traces
+	trace.Event("resolving intents [async]")
+	var wg sync.WaitGroup
+
+	bArgs := &proto.BatchRequest{}
+	bArgs.User = security.RootUser
+	for i := range intents {
+		intent := intents[i] // avoids a race in `i, intent := range ...`
+		var resolveArgs proto.Request
+		var local bool // whether this intent lives on this Range
+		{
+			header := proto.RequestHeader{
+				// Use the pushee's timestamp, which might be lower than the
+				// pusher's request timestamp. No need to push the intent higher
+				// than the pushee's txn!
+				Timestamp: intent.Txn.Timestamp,
+				Key:       intent.Key,
+				EndKey:    intent.EndKey,
+				User:      security.RootUser,
+				Txn:       &intent.Txn,
+			}
+
+			if len(intent.EndKey) == 0 {
+				resolveArgs = &proto.ResolveIntentRequest{RequestHeader: header}
+				local = r.ContainsKey(intent.Key)
+			} else {
+				resolveArgs = &proto.ResolveIntentRangeRequest{RequestHeader: header}
+				local = r.ContainsKeyRange(intent.Key, intent.EndKey)
+			}
+		}
+
+		// If the intent isn't (completely) local, we'll need to send an external request.
+		// We'll batch them all up and send at the end.
+		if !local {
+			bArgs.Add(resolveArgs)
+			continue
+		}
+
+		// If it is local, it goes directly into Raft.
+		wg.Add(1)
+		action := func() {
+			// Trace this under the ID of the intent owner.
+			ctx := tracer.ToCtx(ctx, r.rm.Tracer().NewTrace(resolveArgs.Header().Txn))
+			if _, err := r.addWriteCmd(ctx, resolveArgs, &wg); err != nil && log.V(0) {
```

Comment:
This doesn't seem either necessary or desirable. Why not just call engine.ResolveIntent directly? Calling addWriteCmd unnecessarily threads this through raft again. Keep in mind that all replicas already have the same set of intents to resolve!